File: bindings/dart/lib/src/metadata.dart
import 'dart:ffi';

import 'package:ffi/ffi.dart';

import 'ffi/bindings.dart';

/// Metadata for a scalar attribute in a collection.
class ScalarMetadata {
  final String name;
  final int dataType;
  final bool notNull;
  final bool primaryKey;
  final String? defaultValue;
  final bool isForeignKey;
  final String? referencesCollection;
  final String? referencesColumn;

  const ScalarMetadata({
    required this.name,
    required this.dataType,
    required this.notNull,
    required this.primaryKey,
    required this.defaultValue,
    required this.isForeignKey,
    required this.referencesCollection,
    required this.referencesColumn,
  });

  /// Creates a ScalarMetadata from a native quiver_scalar_metadata_t struct.
  ScalarMetadata.fromNative(quiver_scalar_metadata_t native)
      : name = native.name.cast<Utf8>().toDartString(),
        dataType = native.data_type,
        notNull = native.not_null != 0,
        primaryKey = native.primary_key != 0,
        defaultValue = native.default_value == nullptr
            ? null
            : native.default_value.cast<Utf8>().toDartString(),
        isForeignKey = native.is_foreign_key != 0,
        referencesCollection = native.references_collection == nullptr
            ? null
            : native.references_collection.cast<Utf8>().toDartString(),
        referencesColumn = native.references_column == nullptr
            ? null
            : native.references_column.cast<Utf8>().toDartString();
}

/// Metadata for a vector, set, or time series group in a collection.
class GroupMetadata {
  final String groupName;
  final String dimensionColumn;
  final List<ScalarMetadata> valueColumns;

  const GroupMetadata({
    required this.groupName,
    required this.dimensionColumn,
    required this.valueColumns,
  });

  /// Creates a GroupMetadata from a native quiver_group_metadata_t struct.
  GroupMetadata.fromNative(quiver_group_metadata_t native)
      : groupName = native.group_name.cast<Utf8>().toDartString(),
        dimensionColumn = native.dimension_column == nullptr
            ? ''
            : native.dimension_column.cast<Utf8>().toDartString(),
        valueColumns = List.generate(
          native.value_column_count,
          (i) => ScalarMetadata.fromNative(native.value_columns[i]),
        );
}

File: bindings/python/src/quiverdb/metadata.py
from __future__ import annotations

from dataclasses import dataclass, field
from enum import IntEnum


class DataType(IntEnum):
    """Data type constants matching C API quiver_data_type_t."""

    INTEGER = 0
    FLOAT = 1
    STRING = 2
    DATE_TIME = 3
    NULL = 4


class LogLevel(IntEnum):
    """Console log level constants matching C API quiver_log_level_t."""

    DEBUG = 0
    INFO = 1
    WARN = 2
    ERROR = 3
    OFF = 4


@dataclass(frozen=True)
class CSVOptions:
    """Options for CSV import and export operations."""

    date_time_format: str = ""
    enum_labels: dict[str, dict[str, dict[str, int]]] = field(default_factory=dict)


@dataclass(frozen=True)
class ScalarMetadata:
    """Metadata for a scalar attribute in a collection."""

    name: str
    data_type: DataType
    not_null: bool
    primary_key: bool
    default_value: str | None
    is_foreign_key: bool
    references_collection: str | None
    references_column: str | None


@dataclass(frozen=True)
class GroupMetadata:
    """Metadata for a vector, set, or time series group in a collection."""

    group_name: str
    dimension_column: str  # empty string for vector/set groups
    value_columns: list[ScalarMetadata]


@dataclass(frozen=True)
class UiMetadata:
    """PSR `<db_dir>/ui/` sidecar record for one scalar attribute (Phase 3, META-01/META-02).

    Every string field is a plain `str`, never `str | None` -- D-13 makes empty string the
    spelling of absence on every FFI layer. `configured` is the discriminator between "the
    sidecar declares this" and "nothing declared"; `label` is never back-filled from an id, so a
    declared-but-blank label (`configured is True`, `label == ""`) is distinct from an
    unconfigured attribute (`configured is False`, every string field empty).
    """

    configured: bool
    label: str
    tooltip: str
    unit: str
    format: str
    icon: str
    hidden: bool
    vocabulary: str
    display_order: int


@dataclass(frozen=True)
class UiEnumEntry:
    """One entry in an enum vocabulary (Phase 3, META-05). Returned in declaration order by
    Database.get_ui_vocabulary -- never a null cell, since D-11 crosses the vocabulary as two
    parallel arrays rather than a struct with optional fields."""

    code: int
    label: str
